[GeoMechanicsApplication] Extract plane strain and three dimensional stress state functionality

rfaasse opened the pull request:
**📝 Description**
This PR aims to extract the plane strain and three dimensional stress state functionality, such that the element does not know about the implementation for the different flavours.

This includes:
- Creating the ThreeDimensionalStressState and PlaneStrainStressState with implementations for `CalculateBMatrix`, `CalculateIntegrationCoefficient` and `CalculateGreenLagrangeStrain` and a `Clone` function, needed due to Kratos' registration system.
- Add unit tests to test if the implementations are correct
- Use the calculation functions mentioned above in the elements.

rfaasse committed:
Created the StressStatePolicy Interface, a first implementation of the axisymmetric policy and added a test for calculating the B matrix

rfaasse committed:
Implement functionality to create BMatrix of the correct size

rfaasse committed:
Minor rename of some constants in axisymmetric_stress_state_policy.cpp

rfaasse committed:
Added the actual calculation of the BMatrix.

rfaasse committed:
Formatted the expected matrix

rfaasse committed:
Made the input vector and matrix more complex and revalidated the results

rfaasse committed:
Refactored the calculation

rfaasse committed:
Added a dummy implementation of calculate integration coefficient

rfaasse committed:
Implemented the calculation of the axisymmetric integration coefficient.

rfaasse committed:
Implemented the clone functionality for the axisymmetric stress state policy

rfaasse committed:
Refactoring duplicate setup in tests

rfaasse committed:
Use the new policy in all axisymmetric elements

rfaasse committed:
Removed some code smells in the axisymmetric header files

rfaasse committed:
Merge remote-tracking branch 'origin/master' into geo/12004-extract-stress-state-functionality

rfaasse committed:
Merge branch 'master' into geo/12004-extract-stress-state-functionality

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_axisymmetric_element.hpp

rfaasse committed:
Renamed the implementation class

rfaasse committed:
Renamed variable

rfaasse committed:
Added some comments to explain the expected matrix

rfaasse committed:
Created plane strain stress state and test help utilities

rfaasse committed:
re-used utility function in different test

rfaasse committed:
Added const

rfaasse committed:
Reformatting

rfaasse committed:
Use the new policy in the upw elements

rfaasse committed:
Remove unused calculation of cauchy almansi strain

rfaasse committed:
Added function to stress state for calculating green lagrange strain, which throws for the axisymmetric case

rfaasse committed:
Use the green lagrange calculation in the elements

rfaasse committed:
Rename CauchyGreen strain to GreenLagrange strain

rfaasse committed:
Fix wrong rename

rfaasse committed:
Renamed files to reflect the class name

rfaasse committed:
Fix pipeline build

rfaasse committed:
Fix pipeline build attempt 2

rfaasse committed:
Fix pipeline build attempt 2

rfaasse committed:
Merge branch 'geo/12004-extract-stress-state-functionality' into geo/12004-extract-plane-and-three-d-stress-state

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_axisymmetric_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/stress_state_policy.h
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp

rfaasse committed:
Added empty function for greenlagrange

rfaasse committed:
Fix pipeline attempt 3

rfaasse committed:
Merge remote-tracking branch 'origin/geo/12004-extract-stress-state-functionality' into geo/12004-extract-stress-state-functionality

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_axisymmetric_element.hpp

rfaasse committed:
Merge branch 'geo/12004-extract-stress-state-functionality' into geo/12004-extract-plane-and-three-d-stress-state

indigocoral committed:
Created 3D stress state and added test for it

indigocoral committed:
Fixed the CalculateBMatrix test for 3D

indigocoral committed:
Removed the old policy in the U_Pw elements

indigocoral committed:
Added a first test for calculating Green Lagrange Strain

indigocoral committed:
Removed unused parameters from small_strain_U_Pw_diff_order_element.cpp to fix the Docker GCC_FullDebug error

indigocoral committed:
Moved the logic of CalculateGreenLagrangeStrain to the three_dimension_stress_state and added the test for it

rfaasse committed:
Merge remote-tracking branch 'origin/master' into geo/12004-extract-plane-and-three-d-stress-state

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_axisymmetric_FIC_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_axisymmetric_FIC_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_axisymmetric_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_axisymmetric_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_updated_lagrangian_FIC_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_updated_lagrangian_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/axisymmetric_stress_state.cpp
#	applications/GeoMechanicsApplication/custom_elements/axisymmetric_stress_state.h
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_axisymmetric_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_axisymmetric_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/stress_state_policy.h
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_FIC_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_axisymmetric_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_axisymmetric_element.hpp
#	applications/GeoMechanicsApplication/custom_elements/updated_lagrangian_U_Pw_diff_order_element.cpp
#	applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp

rfaasse committed:
Minor comment change and removal of redundant semi-colon

indigocoral committed:
Merge remote-tracking branch 'refs/remotes/origin/geo/12004-extract-plane-and-three-d-stress-state' into geo/12004-extract-plane-and-three-d-stress-state

indigocoral committed:
Fixed the argument being given to the CalculateGreenLagrangeStrain in Small Strain UPw files, Renamed the class and the file names and did some refactoring

rfaasse committed:
Added calculation of green lagrange strain to the plane strain stress state

rfaasse committed:
Reformatting

rfaasse committed:
Minor clean-up

rfaasse committed:
Extracted utility function for calculating the tensor for green lagrange strain.

Moreover, used std::unique_ptr instead of unique_ptr in clone functions

rfaasse committed:
Made all the if-statements for creating the stress-state policies consistent (to make it easier to see later how to refactor into a member)

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:
```diff
@@ -700,7 +702,15 @@ double UPwBaseElement<TDim, TNumNodes>::CalculateIntegrationCoefficient(
+    }
+
+    return stress_state_policy->CalculateIntegrationCoefficient(IntegrationPoints[PointNumber],
+                                                                detJ, GetGeometry());
```
I know that the calculation of the integration coefficient is the same for plane strain and 3D, so this might seem overkill. However, this makes it consistent with all the other stress-state dependent functions, such that making the stress state policy a member of the elements becomes simpler.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1209,33 +1211,15 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateBMatrix(Matrix& rB, const
-            rB(INDEX_3D_XZ, index + INDEX_Z) = GradNpT(i, INDEX_X);
-        }
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if constexpr (TDim == 2) {
```
This if-statement is there a couple of times for the elements now. This will all disappear when we pass the policy as a member (then we just delegate the call and the type of policy will have been decided upon construction).

avdg81 replied:
Thanks for explaining. Agreed.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+
+Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
@indigocoral I added a utility function to calculate the ETensor, which we now re-use for 3D and plane strain.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.hpp:
```diff
@@ -132,6 +132,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStrainUtilities
         KRATOS_CATCH("")
     }
 
+    static Matrix CalculateGreenLagrangeStrainTensor(const Matrix& rDeformationGradient);
```
Here also all other functions can be split, such that we do not need a .hpp but a cleaner .h and .cpp files. I'd propose to do that in a separate PR.

WPK4FEM replied:
Yes please

avdg81 replied:
Agreed.

rfaasse committed:
Minor clean-up

rfaasse committed:
Cleaning up auto and the order of functions in three_dimensional_stress_state.cpp

rfaasse committed:
Extended the three dimensional stress state test for green lagrange strain to also contain non-diagonal deformation gradient terms

rfaasse committed:
Extended the three dimensional stress state test for green lagrange strain to also contain non-diagonal deformation gradient terms

rfaasse committed:
Merge remote-tracking branch 'origin/geo/12004-extract-plane-and-three-d-stress-state' into geo/12004-extract-plane-and-three-d-stress-state

# Conflicts:
#	applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+    Matrix     result = ZeroMatrix(VOIGT_SIZE_2D_AXISYMMETRIC, dimension * number_of_nodes);
+
+    for (unsigned int i = 0; i < number_of_nodes; ++i) {
+        const auto index = dimension * i;
```
this is more an offset than an index

rfaasse replied:
Good point, changed the naming!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+{
+    const auto dimension       = rGeometry.WorkingSpaceDimension();
+    const auto number_of_nodes = rGeometry.size();
+    Matrix     result = ZeroMatrix(VOIGT_SIZE_2D_AXISYMMETRIC, dimension * number_of_nodes);
```
I expected alignment of the =, but that may have been overruled by the length of the line.

rfaasse replied:
Indeed, aligning it would result in a width of 102 characters unfortunately

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+    Matrix     result          = ZeroMatrix(VOIGT_SIZE_3D, dimension * number_of_nodes);
+
+    for (unsigned int i = 0; i < number_of_nodes; ++i) {
+        auto index = dimension * i;
```
Also this is more an offset than an index.

avdg81 replied:
The variable name has been changed. Now it is also consistent with the plane strain implementation.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_axisymmetric_stress_state.cpp:
```diff
@@ -17,34 +17,21 @@
 
     Model model;
-    auto& model_part = CreateModelPart(model);
+    auto& model_part = ModelSetupUtilities::CreateModelPartWithASingle2D3NElement(model);
```
Should the variable name start with r?

rfaasse replied:
Walked through the axisymmetric and plane strain tests and changed the model_part that I could find to r_model_part. Could you do the same for the 3D tests @indigocoral?

avdg81 replied:
On behalf of @indigocoral it has been done.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.h:
```diff
@@ -0,0 +1,34 @@
+public:
+    [[nodiscard]] Matrix CalculateBMatrix(const Matrix&         GradNpT,
+                                          const Vector&         Np,
+                                          const Geometry<Node>& rGeometry) const override;
```
Does this follow the naming conventions?

avdg81 replied:
Good catch. The parameter names have been updated.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+
+Matrix ThreeDimensionalStressState::CalculateBMatrix(const Matrix&         GradNpT,
+                                                     const Vector&         Np,
+                                                     const Geometry<Node>& rGeometry) const
```
Does this follow the naming conventions?

avdg81 replied:
In addition to Wijtze Pieter's comment, I would suggest to remove the name `Np` here, since it is not being used at all in this override. Please, also remove the parameter name in the corresponding header file.

avdg81 replied:
Both remarks have been addressed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:
```diff
@@ -2278,32 +2264,14 @@ Vector SmallStrainUPwDiffOrderElement::CalculateGreenLagrangeStrain(const Matrix
-        result[INDEX_2D_PLANE_STRAIN_XY] = StrainVector[2];
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if (GetGeometry().WorkingSpaceDimension() == 2) {
+        stress_state_policy = std::make_unique<PlaneStrainStressState>();
```
Should these start with a p?

rfaasse replied:
Agreed, changed all of them to p_stress_state_policy

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+    result[INDEX_2D_PLANE_STRAIN_YY] = StrainVector[1];
+    result[INDEX_2D_PLANE_STRAIN_ZZ] = 0.0;
+    result[INDEX_2D_PLANE_STRAIN_XY] = StrainVector[2];
+
```
I think this part of line 50 to 54 I've seen repeatedly. Should it therefore become something in the stress_strain utitlities.

rfaasse replied:
I agree, but in my opinion the `PlaneStrainStressState` is the only place it should be (converting a strain tensor to a vector). For example, the Hencky strain calculation could also move here in the same way as for Green Lagrange: The calculation of the tensor would be done in StressStrainUtilities and the conversion to the plain strain vector would be done in this class. What is your opinion on this?

WPK4FEM replied:
Indeed, the tensor based part in a utility, then the vector part that is specific for the stress state in the stress state functionality.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:
```diff
@@ -700,7 +702,15 @@ double UPwBaseElement<TDim, TNumNodes>::CalculateIntegrationCoefficient(
-    return IntegrationPoints[PointNumber].Weight() * detJ;
+    std::unique_ptr<StressStatePolicy> stress_state_policy;
+    if constexpr (TDim == 2) {
+        stress_state_policy = std::make_unique<PlaneStrainStressState>();
```
p_stress_state_policy ?

rfaasse replied:
Yes, done

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h:
```diff
@@ -0,0 +1,33 @@
+{
+public:
+    [[nodiscard]] Matrix CalculateBMatrix(const Matrix&         GradNpT,
+                                          const Vector&         Np,
```
rGradNpT and rNp ?

rfaasse replied:
Good catch! Changed

WPK4FEM reviewed (commented):
I looks like this will get rid of a lot of duplicated code and provides tested basic functionality.

rfaasse committed:
Processing review comments for plane strain and axisymmetric stress states and the elements

rfaasse reviewed (commented):
Thanks for the quick review @WPK4FEM! I processed the comments except for the ones related to the 3D stress state, which @indigocoral will pick up.

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+double PlaneStrainStressState::CalculateIntegrationCoefficient(const Geometry<Node>::IntegrationPointType& rIntegrationPoint,
+                                                               double detJ,
```
Nitpicking (to follow the Kratos Style Guide):
```suggestion
                                                               double DetJ,
```

rfaasse replied:
Done

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+
+Vector PlaneStrainStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
Perhaps naming the result `strain_tensor` is clearer than `ETensor`? (And it would be more in line with the variable name in the next statement: `StrainVector`, which should be `strain_vector`.) In fact, since the tensor form is not used (we only use the vector form), we may consider to merge the two statements, i.e.:
```
const auto strain_vector = MathUtils<>::StrainTensorToVector(StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient));
```

WPK4FEM replied:
The symbol E is used for the GreenLagrange strain tensor in literature. So I disagree with Anne here.

rfaasse replied:
I extracted a function that converts the strain tensor to a vector. Since that one can be used later for more than only Green Lagrange (e.g. for Hencky Strain), I made the argument name there `rStrainTensor`. Is this fine for everyone?

avdg81 replied:
Yes, I'm fine with that parameter name.

WPK4FEM replied:
Ok

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.h:
```diff
@@ -0,0 +1,33 @@
+    [[nodiscard]] double CalculateIntegrationCoefficient(const Geometry<Node>::IntegrationPointType& rIntegrationPoint,
+                                                         double detJ,
+                                                         const Geometry<Node>& rGeometry) const override;
+    [[nodiscard]] Vector CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const override;
```
After a brief discussion with Wijtze Pieter, we feel that a better parameter name would be `rDeformationGradient` (i.e. stripping of the `Total` part).

rfaasse replied:
Done!

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,64 @@
+{
+
+Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         rGradNpT,
+                                                const Vector&         rNp,
```
I would suggest to make this parameter unnamed (i.e. remove `rNp`), since it is not being used by this override. Please, also propagate that change to the header file.

rfaasse replied:
Good point, done

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+double ThreeDimensionalStressState::CalculateIntegrationCoefficient(
+    const Geometry<Node>::IntegrationPointType& rIntegrationPoint, double detJ, const Geometry<Node>&) const
```
More nitpicking:
```suggestion
    const Geometry<Node>::IntegrationPointType& rIntegrationPoint, double DetJ, const Geometry<Node>&) const
```

WPK4FEM replied:
Was done

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+    return rIntegrationPoint.Weight() * detJ;
+}
+
+Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
```
As indicated before, please rename this function parameter:
```suggestion
Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rDeformationGradient) const
```

avdg81 commented on applications/GeoMechanicsApplication/custom_elements/three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,61 @@
+
+Vector ThreeDimensionalStressState::CalculateGreenLagrangeStrain(const Matrix& rTotalDeformationGradient) const
+{
+    const auto ETensor = StressStrainUtilities::CalculateGreenLagrangeStrainTensor(rTotalDeformationGradient);
```
See my previous remarks about the name `ETensor` and the possibility to merge this statement with the next one.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_plane_strain_stress_state.cpp:
```diff
@@ -0,0 +1,104 @@
+
+KRATOS_TEST_CASE_IN_SUITE(CalculateBMatrixGivesCorrectResults, KratosGeoMechanicsFastSuite)
+{
+    auto stress_state = std::make_unique<PlaneStrainStressState>();
```
Nitpicking:
```suggestion
    const auto p_stress_state = std::make_unique<PlaneStrainStressState>();
```

rfaasse replied:
Good catch!

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,115 @@
+    const auto p_stress_state_policy = std::make_unique<ThreeDimensionalStressState>();
+
+    Model model;
+    auto& model_part = ModelSetupUtilities::CreateModelPartWithASingle2D3NElement(model);
```
It may be confusing to generate a 2D element for this test, since it aims at a 3D stress state.

WPK4FEM replied:
Element was changed to a 3D element

avdg81 reviewed (commented):
This is a very clear and clean PR. The hard work that you (@rfaasse and @indigocoral) have put into this is much appreciated, and I'm sure it will pay off in the near future. A big thank you! I only have several minor comments, so I believe we can merge this to `master` soon.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_three_dimensional_stress_state.cpp:
```diff
@@ -0,0 +1,115 @@
+
+KRATOS_TEST_CASE_IN_SUITE(ThreeDimensionalStressState_CalculateBMatrixReturnsCorrectResults, KratosGeoMechanicsFastSuite)
+{
+    auto stress_state = std::make_unique<ThreeDimensionalStressState>();
```
Similar to @avdg81's comment for the plane strain test, this one can also be p_stress_state

WPK4FEM replied:
Has become p_stress_state_policy.

rfaasse reviewed (commented):
Processed the comments for the plane/axi states/tests. @indigocoral and I will have a look tomorrow at the comments on 3D.

rfaasse committed:
Processing review comments Anne for plane strain and axisymmetric stress states.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/plane_strain_stress_state.cpp:
```diff
@@ -36,29 +34,31 @@ Matrix PlaneStrainStressState::CalculateBMatrix(const Matrix&         rGradNpT,
     return std::make_unique<PlaneStrainStressState>();
 }
 
+Vector PlaneStrainStressState::ConvertStrainTensorToVector(const Matrix& rStrainTensor)
```
Nice!

avdg81 committed:
Addressed the remaining review comments

- Renamed a few local variables to comply with the Kratos Style Guide.
- Renamed `index` to `offset` to make it consistent across
  implementations.
- Unnamed an unused function parameter.
- Made a local variable `const`.
- Eliminated an intermediate variable that was not strictly needed.
- Use 3D elements when testing the 3D stress state policy.

WPK4FEM reviewed (approved):
Tried to go over the comments 1 by 1 and looked in the last version of the code. I'm happy with the current state.
